File: python/countSort.py

def countSort(arr, M):
  count = [0] * M
  for i in range(len(arr)):
    count[arr[i]] += 1
  
  pos = [0] * M
  for j in range(1, M):
    pos[j] = pos[j - 1] + count[j - 1]
  
  ret = [0] * len(arr)
  for i in range(len(arr)):
    ret[pos[arr[i]]] = arr[i]
    pos[arr[i]] += 1
    
  return ret

arr = [2,3,2,1,3,2,2,3,2,2,2,1]
print(countSort(arr, 4))

File: java/CountSort.java
package algo;

public class CountSort {

	private static int[] countSort(int[] arr, int M) {
		int[] count = new int[M];
		for (int i = 0 ; i < arr.length ; i++)
			count[arr[i]] += 1;
		
		int[] pos = new int[M];
		for (int j = 1 ; j < M ; j++)
			pos[j] = pos[j - 1] + count[j - 1];
		
		int[] ret = new int[arr.length];
		for (int i = 0 ; i < arr.length ; i++) {
			ret[pos[arr[i]]] = arr[i];
			pos[arr[i]] += 1;
		}
		
		return ret;
	}
	
	public static void main(String[] args) {
		int[] arr = {2,3,2,1,3,2,2,3,2,2,2,1};
		for (int i : countSort(arr, 4)) {
			System.out.print(i +  ",");
		}
	}
}
